tests: add worker label through go

bookinabox opened the pull request:
libvirt e2e tests did not properly label worker nodes with the worker label.

This creates a function in test/provisioner/common.go that adds the worker label to the libvirt e2e createCluster.

Fixes: #1107

wainersm commented:
Hi @bookinabox !

You will need to fix the error found by the golangci-lint:

```
./hack/golangci-lint.sh
Error: test/provisioner/common.go:26:14: ineffectual assignment to err (ineffassign)
		clientset, err = kubernetes.NewForConfig(config)
		           ^
make: *** [Makefile:113: golangci-lint] Error 1
```

wainersm commented:
I am also getting an error:

```
Adding worker label to nodes belonging to: peer-pods
W0710 15:07:13.356706   29753 client_config.go:618] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
W0710 15:07:13.356717   29753 client_config.go:623] error creating inClusterConfig, falling back to default config: unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined
E0710 15:07:13.356787   29753 env.go:352] Recovering from panic and running finish actionsruntime error: invalid memory address or nil pointer dereference
```

I don't have `KUBECONFIG` exported on my environment. I let the e2e framework to create the cluster for me.

bookinabox commented:
> I am also getting an error:
> 
> ```
> Adding worker label to nodes belonging to: peer-pods
> W0710 15:07:13.356706   29753 client_config.go:618] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
> W0710 15:07:13.356717   29753 client_config.go:623] error creating inClusterConfig, falling back to default config: unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined
> E0710 15:07:13.356787   29753 env.go:352] Recovering from panic and running finish actionsruntime error: invalid memory address or nil pointer dereference
> ```
> 
> I don't have `KUBECONFIG` exported on my environment. I let the e2e framework to create the cluster for me.

Going to source the kcli_cluster.sh file then, so that the export has an effect outside the script (it doesn't look like its doing anything otherwise). I could also hard code the same path in, but I think to keep everything consistent, that might be best to only have the path declared in one place.

katexochen commented on test/provisioner/common.go:
```diff
@@ -0,0 +1,64 @@
+	}
+
+	fmt.Printf("initialized variables\n")
+	nodelist, err := clientset.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+
+	if err != nil {
+		return err
+	}
```
```suggestion
	nodelist, err := clientset.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
	if err != nil {
		return err
	}
```

katexochen commented on test/provisioner/provision_libvirt.go:
```diff
@@ -100,6 +101,16 @@ func (l *LibvirtProvisioner) CreateCluster(ctx context.Context, cfg *envconf.Con
+		return err
+	}
+
+	err = AddNodeRoleWorkerLabel(clusterName, cfg)
+
+	if err != nil {
+		return fmt.Errorf("labeling nodes: %w", err)
+	}
```
```suggestion
	if err := AddNodeRoleWorkerLabel(clusterName, cfg); err != nil {
		return fmt.Errorf("labeling nodes: %w", err)
	}
```

katexochen commented on test/provisioner/provision_libvirt.go:
```diff
@@ -100,6 +101,16 @@ func (l *LibvirtProvisioner) CreateCluster(ctx context.Context, cfg *envconf.Con
 	kubeconfig := path.Join(home, ".kcli/clusters", clusterName, "auth/kubeconfig")
 	cfg.WithKubeconfigFile(kubeconfig)
 
+	if err != nil {
+		return err
+	}
+
```
What error are we checking here?

bookinabox replied:
oops, overlooked that. Thanks!

katexochen commented on test/provisioner/common.go:
```diff
@@ -0,0 +1,64 @@
+		return err
+	}
+
+	label := map[string]string{"node-role.kubernetes.io/worker": ""}
```
Depending on merge order, this might needs to updated to fit the changes in https://github.com/confidential-containers/cloud-api-adaptor/pull/784

bookinabox commented:
updated to new worker label

katexochen reviewed (commented):
Sorry, missed this one on the first review. Thanks for updating to the new label.

bookinabox commented on test/provisioner/provision_libvirt.go:
```diff
@@ -100,6 +101,11 @@ func (l *LibvirtProvisioner) CreateCluster(ctx context.Context, cfg *envconf.Con
 	kubeconfig := path.Join(home, ".kcli/clusters", clusterName, "auth/kubeconfig")
 	cfg.WithKubeconfigFile(kubeconfig)
 
+	if err := AddNodeRoleWorkerLabel(clusterName, cfg); err != nil {
```
Updated!

wainersm commented on test/provisioner/common.go:
```diff
@@ -0,0 +1,62 @@
+func AddNodeRoleWorkerLabel(ctx context.Context, clusterName string, cfg *envconf.Config) error {
+	fmt.Printf("Adding worker label to nodes belonging to: %s\n", clusterName)
+
+	config, err := clientcmd.BuildConfigFromFlags("", cfg.KubeconfigFile())
```
@bookinabox I tested this PR, it is working. I've only a few comments to the code.

First, what if you get the client object like (https://pkg.go.dev/sigs.k8s.io/e2e-framework@v0.1.0/pkg/envconf#Config.NewClient):
```
	client, err := cfg.NewClient()
	if err != nil {
		return err
	}
```

bookinabox replied:
Well, the clientset and the client returned from that function are different types. I don't know why there are different ways to do the same thing/what the pros/cons are but I will try to do it this way. I imagine there will be some major changes required to do this though.

bookinabox replied:
Is there a specific reason why this way would be better? I'm having a little trouble getting the patch operation to work with resources (I'm sure there's an easy fix, I'm confused which k8s.object I need to pass into .Patch) and I'm not sure what the benefit to doing the above way would be.

wainersm replied:
I saw that you managed to use the cfg.NewClient() which returns a `kclient` (https://github.com/kubernetes-sigs/e2e-framework/tree/main/klient), however I still own you some answers.

The e2e-framework provides some facilities over the standard k8s api. For example, when you start the test suite you can set the k8s namespace resources will be created (and destroyed on the end). The `kcliet` wraps that information so that you don't need to pass the namespace to every call of, for example, `Resources()`. The `kclient` has also special `wait` routines for resources. And so on.

There are other reasons to use 'kclient' instead of 'clientset`. For example, avoid dependence on another package, avoid code fragmentation by leveraging a single API thoughtout the code base. And so on.

wainersm commented on test/provisioner/common.go:
```diff
@@ -0,0 +1,62 @@
+	Value map[string]string `json:"value"`
+}
+
+func AddNodeRoleWorkerLabel(ctx context.Context, clusterName string, cfg *envconf.Config) error {
```
did you try to extract the cluster name from the Config and/or Client objects? If it is possible, it would save us one argument to this function :)

wainersm replied:
While talking about this function, as it will have public access...could you please add documentation?

wainersm commented on test/provisioner/common.go:
```diff
@@ -0,0 +1,62 @@
+package provisioner
```
Could you add a license header to this file?

See https://github.com/confidential-containers/cloud-api-adaptor/blob/main/test/provisioner/provision.go as for example:
```
// (C) Copyright Confidential Containers Contributors
// SPDX-License-Identifier: Apache-2.0
```

wainersm commented on test/provisioner/common.go:
```diff
@@ -0,0 +1,62 @@
+
+	for _, node := range nodelist.Items {
+		if strings.Contains(node.Name, workerStr) {
+			_, err = clientset.CoreV1().Nodes().Patch(ctx, node.Name, types.JSONPatchType, payloadBytes, metav1.PatchOptions{})
```
If the label already exists, will it fail? If yes then maybe it should check its existence first.

bookinabox replied:
afaik, it shouldn't fail, at least with this API. It has not errored out for that reason when I tested it.

wainersm commented on test/provisioner/common.go:
```diff
@@ -0,0 +1,62 @@
+		if strings.Contains(node.Name, workerStr) {
+			_, err = clientset.CoreV1().Nodes().Patch(ctx, node.Name, types.JSONPatchType, payloadBytes, metav1.PatchOptions{})
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Unable to add label %s to node %s: %s\n", label, node.Name, err)
```
are you not returning in case of error on purpose?

bookinabox replied:
I think I was at one point, but its probably better to error out.

wainersm commented on test/provisioner/provision_libvirt.go:
```diff
@@ -100,6 +101,11 @@ func (l *LibvirtProvisioner) CreateCluster(ctx context.Context, cfg *envconf.Con
 	cfg.WithKubeconfigFile(kubeconfig)
 
+	if err := AddNodeRoleWorkerLabel(ctx, clusterName, cfg); err != nil {
+
```
don't need this extra newline.

bookinabox commented:
I changed it to use the different API and cleaned up the code a bit. One thing I'm still looking into is the cluster name stuff. It looks like kcli creates a cluster with the generic "kubernetes" cluster name. I haven't looked deep enough into the code to determine how it figures out how to delete them, but I will investigate.

This makes the "clustername" argument misleading, so I'm thinking of a better way/how to work around this.

wainersm commented:
> I changed it to use the different API and cleaned up the code a bit. One thing I'm still looking into is the cluster name stuff. It looks like kcli creates a cluster with the generic "kubernetes" cluster name. I haven't looked deep enough into the code to determine how it figures out how to delete them, but I will investigate.
> 
> This makes the "clustername" argument misleading, so I'm thinking of a better way/how to work around this.

Do you prefer to merge as is and you send another PR later? IMO let's try to merge this because it will ease the testing of the other pending PRs....

wainersm commented:
I tested the latest version of this PR, working fine on my env.

bookinabox commented:
okay, let's just merge it for now then.

wainersm commented:
hi @katexochen ! Do you have more comments for this PR?

katexochen commented:
Sorry, I didn't find the time to debug my local setup, but code lgtm. :+1:

snir911 reviewed (approved):
LGTM thanks!
